core-rust/state-manager/src/store/db.rs

```diff
@@ -87,6 +87,32 @@ use crate::{
 use radix_engine::types::{KeyValueStoreId, SubstateId};
 use radix_engine_stores::hash_tree::tree_store::{NodeKey, Payload, ReadableTreeStore, TreeNode};
 
+// TODO(engine-merge): remove (use the impl in radix engine)
+pub fn encode_substate_id(index_id: &Vec<u8>, db_key: &Vec<u8>) -> Vec<u8> {
+    let mut buffer = Vec::new();
+    buffer.extend(index_id);
+    buffer.extend(db_key); // Length is marked by EOF
+    buffer
+}
+
+// TODO(engine-merge): remove (use the impl in radix engine)
+pub fn decode_substate_id(slice: &[u8]) -> Option<(Vec<u8>, Vec<u8>)> {
+    if slice.len() >= 26 {
+        let index_id = slice[0..26].to_vec();
+        let key = slice[26 + 1..].to_vec();
+        return Some((index_id, key));
+    }
+    None
+}
+
+// TODO(engine-merge): placeholder trait; remove once radix engine is merged
+pub trait ListableSubstateStore {
+    fn list_substates(
+        &self,
+        index_id: &Vec<u8>,
+    ) -> Box<dyn Iterator<Item = (Vec<u8>, Vec<u8>)> + '_>;
```

Comment:
Is `dyn` necessary here?

Reply:
Either you use Box<dyn Iterator... or you make the iterator an associated type of the trait. Happy with either.

Reply:
I'm in favour of a GAT (compile-time vs runtime).